### ISSUE TITLE
Figure out how to implement more actions

### ISSUE THREAD

**Issue by jonc125** [UCL-CloudLabs/user-portal#9]
There's some info I don't think we can get from the Terraform state:
* Disk type
* Disk size
* Machine status (running/starting/stopped)

Some of these may require using the Azure API?

Other questions:
* How do we start/stop/restart a VM?
* What kinds of edit operation do we want to support? That's probably not needed for the initial demo though.
* How do we download a machine?

Broken-down list of sub-tasks in priority order:
- [x] Display current host status on the Dashboard
- [x] Implement actions to start/stop/restart a VM
- [ ] Implement the 'Download VM' action *(nice to have)*
- [ ] Put hardcoded values for disk type & size in the info page, corresponding to Standard_A0
- [ ] Display the 'last started' time in the info page

Further refinements will be left for future issues.

**Comment by raquelalegre** [comment]
Disk type and size are linked to the machine selected by the user from the free tier. In the example I have hardcoded Standard_A0. This has a specific disk type and size, so we'll need to save that info somewhere or see if the Azure API gives us that info.

Machine status is going to need to be with the  Azure API as well. It can also be started/stoped/restarted like this: 

https://github.com/Azure-Samples/virtual-machines-python-manage/blob/bf5afe9600fb2c820c6d9af43e3817b4c467f145/unmanaged-disks/example.py#L187

Edit operations... maybe SSH keys changing, URL prefixes, username/passwd, description, ... I guess things that are free - i.e. not the disk size for example.

There is a way to download the machine from the portal which I can't see right now because it's not loading the VMs! However it is doable - although they're using the Powershell CLI in the example. I'll try again later and see if the portal is working any better...



https://www.petri.com/download-azure-vm-hyper-v

**Comment by raquelalegre** [comment]
Confirmed there is no way to start/stop/restart a VM on Terraform, so we'll have to use the Azure REST API or the Python API.

**Comment by raquelalegre** [comment]
About the edit options I said "things that are free of charge", but I think it should be more "things that don't imply deploying things on a different machine". That'd imply deploying a new host and wouldn't be enough with editing an existing one.

Maybe you have already thought about this, but we should think about saving configurations to easily reuse in new deployments. For example, if a user deploys a website and later on he realises the disk size is too small, they'd need to use a different machine or an storage system, but the GH repo/s, the SSH keys, the provisioning script should be the same.

**Comment by jonc125** [comment]
Yeah, 'create new machine similar to this one' would be nice. I'll create a new issue for it.

**Comment by jonc125** [comment]
For status, the tests (see `test_deployer` function) currently 'ping' the service to see if it's live. We could do the same in an AJAX request. But we might also want to use the Azure API if there's no response to determine whether the machine is up but service not responding, or machine is stopped, or machine is restarting, etc.

**Comment by jonc125** [comment]
Updating a user's hosts' statuses can now (following #7) be done in a Celery task (or indeed task per host, depending on how the Azure SDK does this) which just updates the DB, and the dashboard could auto-refresh to get the latest statuses from the DB.

**Comment by ageorgou** [comment]
Getting the status is doable, if not completely straightforward. Some information can be found here:

https://docs.microsoft.com/en-us/rest/api/compute/virtualmachines/virtualmachines-state
